google/cloud/internal/tracing_rest_client.cc

```diff
@@ -57,151 +58,160 @@ void InjectCloudTraceContext(RestContext& ctx,
-        static_cast<std::int64_t>(context.appconnect_time()->count()));
-  }
-
-  if (!request_result) return;
-  auto const& response = *request_result;
-  if (!response) return;
   for (auto const& kv : context.headers()) {
     auto const name = "http.request.header." + kv.first;
     if (kv.second.empty()) {
-      span.SetAttribute(name, "");
+      span->SetAttribute(name, "");
       continue;
     }
     if (absl::EqualsIgnoreCase(kv.first, "authorization")) {
-      span.SetAttribute(name, kv.second.front().substr(0, 32));
+      span->SetAttribute(name, kv.second.front().substr(0, 32));
       continue;
     }
-    span.SetAttribute(name, kv.second.front());
+    span->SetAttribute(name, kv.second.front());
+  }
+  if (!request_result || !(*request_result)) {
+    return internal::EndSpan(*span, std::move(request_result));
   }
+  auto const& response = *request_result;
 
   // There are only 32 attributes available per span, and excess attributes are
   // discarded. First add the `x-*` headers. They tend to have more important
   // information.
   for (auto const& kv : response->Headers()) {
     if (!absl::StartsWith(kv.first, "x-")) continue;
-    span.SetAttribute("http.response.header." + kv.first, kv.second);
+    span->SetAttribute("http.response.header." + kv.first, kv.second);
   }
   // Then add all other headers.
   for (auto const& kv : response->Headers()) {
     if (absl::StartsWith(kv.first, "x-")) continue;
-    span.SetAttribute("http.response.header." + kv.first, kv.second);
+    span->SetAttribute("http.response.header." + kv.first, kv.second);
   }
+  return std::unique_ptr<RestResponse>(std::make_unique<TracingRestResponse>(
+      *std::move(request_result), std::move(span)));
 }
 
-/**
- * Extracts information from @p value, and adds it to a span.
- *
- * The span is ended. The original value is returned, for the sake of
- * composition.
- *
- * Note that this function should be called after the RPC has finished.
- */
-StatusOr<std::unique_ptr<RestResponse>> EndResponseSpan(
-    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span,
-    RestContext& context, StatusOr<std::unique_ptr<RestResponse>> value) {
-  if (!value) return internal::EndSpan(*span, std::move(value));
-  ExtractAttributes(context, value, *span);
-  return std::unique_ptr<RestResponse>(std::make_unique<TracingRestResponse>(
-      *std::move(value), std::move(span)));
+opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> HttpStart(
+    std::chrono::system_clock::time_point start, Options const& options) {
+  opentelemetry::trace::StartSpanOptions span_options;
+  span_options.kind = opentelemetry::trace::SpanKind::kClient;
+  span_options.start_system_time = start;
+  return internal::GetTracer(options)->StartSpan("SendRequest", span_options);
+}
+
+StatusOr<std::unique_ptr<RestResponse>> EndStartSpan(
+    opentelemetry::trace::Span& span,
+    std::chrono::system_clock::time_point start, RestContext& context,
```

Comment:
Drive by: We want to do the time math using `absl::Time` to avoid rounding.

As @devbww noticed, `absl::Time` and `opentelemetry::common::SystemTimestamp` have nanosecond resolution, but `std::chrono::system_clock` may not have nanosecond resolution on all platforms.

Reply:
I don't think it can possibly make a difference:

- I am adding microseconds, in all platforms we support `std::chrono::system_clock` has higher resolution than one microsecond.
- `opentelemetry::common::SystemTimestamp` is initialized from a `std::chrono::system_clock::now()`.  Whatever rounding happens, it happens regardless of whether I do the math or the opentelemetry library does the math.

Unless I hear of an specific case in a real computer where this makes a difference, the code as-is is cleaner than going via `absl::Time`.